### PR TITLE
Standardize errors

### PR DESCRIPTION
**Pull request by peterbroadhurst**
- Use a consistent utility to construct errors
- Keep all of the error messages in a single location (Go code for now, extensible to i18n translation)

### CODE REVIEW

**Comment by peterbroadhurst** [comment]
Need to investigate the coverage before this merges

File: internal/klderrors/errors.go [hunk 1]
```diff
@@ -0,0 +1,422 @@
...
+	AddressBookLookupBadURL = "Invalid URL obtained for address"
+	// AddressBookLookupBadHostsFile we have a custom hosts file for DNS resolution, but it cannot be processed
+	AddressBookLookupBadHostsFile = "Configuration problem (hosts file)"
+	// AddressBookLookupNotFound remote addressbook says no
+	AddressBookLookupNotFound = "Unknown address"
+
+	// ConfigFileReadFailed failed to read the server config file
+	ConfigFileReadFailed = "Failed to read %s: %s"
+	// CompilerVerionNotFound the runtime context of ethconnect has not been configured with a compiler for the requested version
+	CompilerVerionNotFound = "Could not find a configured compiler for requested Solidity major version %s.%s"
```
Comment (vdamle) [inline internal/klderrors/errors.go]:
```suggestion
	CompilerVersionNotFound = "Could not find a configured compiler for requested Solidity major version %s.%s"
```

File: internal/klderrors/errors.go [hunk 2]
```diff
@@ -0,0 +1,422 @@
...
+	// CompilerVerionBadRequest the user requested a bad semver
+	CompilerVerionBadRequest = "Invalid Solidity version requested for compiler. Ensure the string starts with two dot separated numbers, such as 0.5"
```
Comment (vdamle) [inline internal/klderrors/errors.go]:
```suggestion
	CompilerVersionBadRequest = "Invalid Solidity version requested for compiler. Ensure the string starts with two dot separated numbers, such as 0.5"
```

File: internal/klderrors/errors.go [hunk 3]
```diff
@@ -0,0 +1,422 @@
...
+	// EventStreamsDBLoad failed to init DB
+	EventStreamsDBLoad = "Failed to open DB at %s: %s"
+	// EventStreamsNoID attempt to create an event stream/sub without an ID
+	EventStreamsNoID = "No ID"
+	// EventStreamsWebhookNoURL attempt to create a Webhook event stream without a URL
+	EventStreamsWebhookNoURL = "Must specify webhook.url for action type 'webhook'"
+	// EventStreamsWebhookInvalidURL attempt to create a Webhook event stream with an invalid URL
+	EventStreamsWebhookInvalidURL = "Invalid URL in webhook action"
+	// EventStreamsWebhookInvalidActionType unknown action type
+	EventStreamsWebhookInvalidActionType = "Unknown action type '%s'"
```
Comment (vdamle) [inline internal/klderrors/errors.go]:
```suggestion
	EventStreamsWebhookInvalidActionType = "Unknown webhook action type '%s'"
```

File: internal/klderrors/errors.go [hunk 4]
```diff
@@ -0,0 +1,422 @@
...
+	KVStoreDBLoad = "Failed to open DB at %s: %s"
+	// KVStoreMemFilteringUnsupported memory db is really just for testing. No filtering support
+	KVStoreMemFilteringUnsupported = "Memory receipts do not support filtering"
+
+	// HDWalletSigningFailed problem returned from remote HDWallet API
+	HDWalletSigningFailed = "HDWallet signing failed"
+	// HDWalletSigningBadData we got a response, but not with the correct fields
+	HDWalletSigningBadData = "Unexpected response from HDWallet"
+	// HDWalletSigningNoConfog we had a request for HD Wallet signing, but we don't have the required config
+	HDWalletSigningNoConfog = "No HD Wallet Configuration"
```
Comment (vdamle) [inline internal/klderrors/errors.go]:
```suggestion
	HDWalletSigningNoConfig = "No HD Wallet Configuration"
```

File: internal/klderrors/errors.go [hunk 5]
```diff
@@ -0,0 +1,422 @@
...
+	// RemoteRegistryNotConfigured cannot register as a remote registry is not configured
+	RemoteRegistryNotConfigured = "No remote registry is configured"
+	// RemoteRegistryRegistrationFailed error during registration with remote contract registry
+	RemoteRegistryRegistrationFailed = "Failed to register instance in remote registry: %s"
+	// RemoteRegistryLookupGatewayNotFound did not find the requested ID in the remote registry for a gateway/factory
+	RemoteRegistryLookupGatewayNotFound = "Gateway not found"
+	// RemoteRegistryLookupInstanceNotFound did not find the requested ID in the remote registry for a contract instance
+	RemoteRegistryLookupInstanceNotFound = "Instance not found"
+	// RemoteRegistryLookupDecodeABIFailed couldn't parse the ABI in the response
+	RemoteRegistryLookupDecodeABIFailed = "Error processing contract registry response"
```
Comment (vdamle) [inline internal/klderrors/errors.go]:
```suggestion
	RemoteRegistryLookupDecodeABIFailed = "Failed to decode ABI: %s"
```

File: internal/kldcontracts/remoteregistry.go [hunk 6]
```diff
@@ -148,7 +148,7 @@ func (rr *remoteRegistry) loadFactoryFromURL(baseURL, ns, lookupStr string) (*de
 	err = json.Unmarshal([]byte(abiString), &abi)
 	if err != nil {
 		log.Errorf("GET %s <-- !Failed to decode ABI: %s\n%s", queryURL, err, abiString)
-		return nil, fmt.Errorf(genericRegistryResponseErrorMsg)
+		return nil, klderrors.Errorf(klderrors.RemoteRegistryLookupDecodeABIFailed)
```
Comment (vdamle) [inline internal/kldcontracts/remoteregistry.go]:
```suggestion
		return nil, klderrors.Errorf(klderrors.RemoteRegistryLookupDecodeABIFailed, err)
```

File: internal/kldcontracts/remoteregistry.go [hunk 7]
```diff
@@ -161,7 +161,7 @@ func (rr *remoteRegistry) loadFactoryFromURL(baseURL, ns, lookupStr string) (*de
 	var bytecode []byte
 	if bytecode, err = hex.DecodeString(strings.TrimPrefix(bytecodeStr, "0x")); err != nil {
 		log.Errorf("GET %s <-- !Failed to parse bytecode: %s\n%s", queryURL, err, bytecodeStr)
-		return nil, fmt.Errorf(genericRegistryResponseErrorMsg)
+		return nil, klderrors.Errorf(klderrors.RemoteRegistryLookupDecodeBytecodeFailed)
```
Comment (vdamle) [inline internal/kldcontracts/remoteregistry.go]:
```suggestion
		return nil, klderrors.Errorf(klderrors.RemoteRegistryLookupDecodeBytecodeFailed, err)
```

File: internal/klderrors/errors.go [hunk 8]
```diff
@@ -0,0 +1,422 @@
...
+	// RemoteRegistryLookupDecodeBytecodeFailed couldn't parse the ABI in the response
+	RemoteRegistryLookupDecodeBytecodeFailed = "Error processing contract registry response"
```
Comment (vdamle) [inline internal/klderrors/errors.go]:
```suggestion
	RemoteRegistryLookupDecodeBytecodeFailed = "Failed to parse bytecode: %s"
```

File: internal/klderrors/errors.go [hunk 9]
```diff
@@ -0,0 +1,422 @@
...
+	RESTGatewaySyncMsgTypeMismatch = "Unexpected condition (message types do not match when processing)"
+	// RESTGatewaySyncWrapErrorWithTXDetail wraps a low level error with transaction hash context on sync APIs before returning
+	RESTGatewaySyncWrapErrorWithTXDetail = "TX %s: %s"
+
+	// RESTGatewayCompileContractInvalidFormData invalid form data when requesting a compilation to generate an ABI/bytecode
+	RESTGatewayCompileContractInvalidFormData = "Could not parse supplied multi-part form data: %s"
+	// RESTGatewayCompileContractCompileFailed failed to perform compile
+	RESTGatewayCompileContractCompileFailed = "Failed to compile solidity: %s"
+	// RESTGatewayCompileContractPostcompileFailed failed to process output of compilation
+	RESTGatewayCompileContractPostcompileFailed = "Failed to process solidity: %s"
```
Comment (vdamle) [inline internal/klderrors/errors.go]:
```suggestion
	RESTGatewayCompileContractPostCompileFailed = "Failed to process solidity: %s"
```

**Comment by peterbroadhurst** [comment]
I've reviewed the coverage @vdamle - it's highlighting cases where we didn't have coverage of error paths previously. I checked the message codes in each of the uncovered lines as I was going through.

I don't think we need to block a merge on this one for the coverage, so once you've had a chance to review I can do the merge.

File: internal/klderrors/errors.go [hunk 10]
```diff
@@ -0,0 +1,422 @@
...
+	// TransactionSendCallFailedRevertMessage directly passes the revert message from the EVM
+	TransactionSendCallFailedRevertMessage = "%s"
+	// TransactionSendCallFailedRevertNoMessage when we couldn't process the EVM revert message
+	TransactionSendCallFailedRevertNoMessage = "EVM reverted. Failed to decode error message"
+	// TransactionSendMissingPrivateFromOrion there is no default privateFrom in Orion, so the user must always supply it
+	TransactionSendMissingPrivateFromOrion = "private-from is required when submitting private transactions via Orion"
+	// TransactionSendPrivateTXWithExternalSigner we don't allow private transactions to be combined with a HD Wallet or other external signer currently
+	TransactionSendPrivateTXWithExternalSigner = "Signing with %s is not currently supported with private transactions"
+	// TransactionSendPrivateForAndPrivactyGroup mixed both params
+	TransactionSendPrivateForAndPrivactyGroup = "privacyGroupId and privateFor are mutually exclusive"
```
Comment (vdamle) [inline internal/klderrors/errors.go]:
```suggestion
	TransactionSendPrivateForAndPrivacyGroup = "privacyGroupId and privateFor are mutually exclusive"
```

File: internal/klderrors/errors.go (hunk 3 above)
Comment (peterbroadhurst) [reply internal/klderrors/errors.go]: Going to remove the webhook from the name on this one, as the intention is this is where we'll put `websockets` / `socket.io` support as an alternative to webhooks

File: internal/kldcontracts/remoteregistry.go (hunk 6 above)
Comment (peterbroadhurst) [reply internal/kldcontracts/remoteregistry.go]: So far, we've deliberately not returned the response over the REST API on this one. So the 'generic' error is intentionally retained

Comment (peterbroadhurst) [reply internal/kldcontracts/remoteregistry.go]: going to rename to avoid confusion though

File: internal/kldcontracts/remoteregistry.go (hunk 7 above)
Comment (peterbroadhurst) [reply internal/kldcontracts/remoteregistry.go]: as above

File: internal/klderrors/errors.go (hunk 8 above)
Comment (peterbroadhurst) [reply internal/klderrors/errors.go]: as above

**Comment by peterbroadhurst** [comment]
Thanks for the great review @vdamle 👍 
There were a couple where I didn't add in the extra insert, as the code intentionally left it out.
Otherwise I've worked through all the corrections.

**Pull request merged**